### PR TITLE
IntersectionTypeTree: Issue #573 

### LINKED ISSUE #573: Lambdas: TypeFromTree.defaultAction: conversion undefined for tree type INTERSECTION_TYPE
[MyChronology.java](https://gist.github.com/dbrosoft/594b3a377435e683cf77) is the smallest example I have that produces the above message; strangely, taking any of the definitions out eliminates the error.  Compilation (using the CF compiler) succeeds, but nullness checking crashes.

### PR DESCRIPTION
**Pull request by smillst**
IntersectionTypeTrees should be handled like other TypeTrees.  This fixes #573.  Also fixes the crash in #478 that occurred when caching was turned off.   

Below is a longer explanation of why caching was masking this bug:
The type of the cast type tree, TypeCastTree.getType(), was cached when creating the type of the cast tree, TypeCastTree. (See TypeFromExpressionVisitory#visitTypeCast.)  So, when AnnotatedTypeFactory#getFunctionalInterfaceType called getAnnotatedType on the cast type tree, the cached expression was returned.   If no cached expression was present, the getAnnotatedType called fromExpression on the cast type tree, but the implementation was missing and thus a crash occurred.

### CODE REVIEW

File: checker/tests/nullness/java8/lambda/Issue573FullTest.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class Issue573FullTest implements Chronology {
```
Comment (smillst) [inline checker/tests/nullness/java8/lambda/Issue573FullTest.java]: It's the test case from the issue.  The reference is at the top of the file.

Comment (wmdietl) [inline checker/tests/nullness/java8/lambda/Issue573FullTest.java]: Yes, but that class looks very much like:
http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f0851bc0e7bf/src/share/classes/java/time/chrono/Chronology.java
which has a big fat Oracle copyright on top.

Comment (smillst) [inline checker/tests/nullness/java8/lambda/Issue573FullTest.java]: Good point. I'll either copy the whole file including the copyright or just remove this test case.

**Comment by smillst** [comment]
Merged into mergeAfterAprilRelease

**Comment by smillst** [comment]
Rebased version merged here: 3727a87

**Comment by wmdietl** [comment]
The test fails with OIGJ. I suppressed the test for now:
https://github.com/typetools/checker-framework/commit/b526d3b7d892cb23aa97141bf0777b427e219623
Please have a look whether there is something better to do.

**Pull request closed without merging**